Change scroll top button to component

yucheng11122017 opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [X] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Fixes #2081

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Change scroll to top button to be a Vue component. This is to be inserted into layout file.
This allow users to 
- Not use it if they want 
- Change the icon 
- Change the icon size and color
- Change the position

**Anything you'd like to highlight/discuss:**

**Testing instructions:**
Insert ```<scroll-top-button></scroll-top-button>``` to layout file. 

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Change scroll top button to component

It is currently added to all pages, cannot be disabled, and does not support styling or positioning. 

This limits the usability and customization for users. 

Changing the scroll top button to a component allows users to not include it in their layout if they want and allows them to style it by passing attributes to it. 

Let's make 
- scroll top button no longer compulsory and to be added to layout file if users want it
- scroll top button added to default file when project is initialized
- support different icons for button 
- support positioning for scroll top button 
- support customized icon size and color

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [X] Updated the documentation for feature additions and enhancements
- [X] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+    },
+    iconStyle() {
+      const style = {};
+      if (this.visible) {
+        style.display = 'block';
+      } else {
+        style.display = 'none';
+      }
```
```suggestion
      style.display = this.visible ? 'block' : 'none';
```

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+      if (this.iconColor) {
+        style.color = this.iconColor;
+      }
+      if (this.bottom) {
+        style.bottom = this.bottom;
+      } else {
+        style.bottom = '2%';
+      }
+      if (this.right) {
+        style.right = this.right;
+      } else {
+        style.right = '2%';
+      }
```
```suggestion
      style.bottom = this.bottom || '2%';
      style.right = this.right || '2%';
```

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+};
+</script>
+
+<style scoped>
```
We can remove this if there are no styles used!

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+      document.body.scrollIntoView({ block: 'start', behavior: 'smooth' });
+    },
+    getIconSize() {
+      let iconSize = '';
+      if (this.iconSize) {
+        iconSize += `fa-${this.iconSize}`;
+      }
+      return iconSize;
```
```suggestion
      return this.iconSize ? `fa-${this.iconSize}` : '';
```

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+<template>
+  <i
+    id="scroll-top-button"
```
I'm a bit unsure about having a fixed ID here - IDs should be unique across a HTML document so if a user defines 2 scroll-to-top buttons (unlikely, but possible) we will have duplicated IDs. Is it possible to avoid this duplication?

We do have some duplicated IDs in our docs as a result of `includes` but it's probably better avoid that :p

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,126 @@
+  },
+  data() {
+    return {
+      visible: false,
+      lighten: false,
```
Minor nit, can we follow the CS2103 convention of having verbs in front of bool variables? e.g. `isVisible`

jovyntls reviewed (changes requested):
Thank you for the work @yucheng11122017 ! This is a nice component to have 🎉 The fading in/out is a nice touch!

I'm not sure if I missed it (from all of the changes in the test files oops) - are there tests for this new component?

yucheng11122017 commented:
> Thank you for the work @yucheng11122017 ! This is a nice component to have 🎉 The fading in/out is a nice touch!
> 
> I'm not sure if I missed it (from all of the changes in the test files oops) - are there tests for this new component?

Hi @jovyntls, thank you for your comments! I have updated them accordingly. I also added test cases for the component. 
One thing I am struggling with is testing that the button appears after the user scrolls. I think the problem is that there is a set timeout which only updates the button after 0.1s so the display in the button doesn't update immediately. I was wondering if you would have any advice on how to tackle this problem? :)

jovyntls commented:
> Hi @jovyntls, thank you for your comments! I have updated them accordingly. I also added test cases for the component. One thing I am struggling with is testing that the button appears after the user scrolls. I think the problem is that there is a set timeout which only updates the button after 0.1s so the display in the button doesn't update immediately. I was wondering if you would have any advice on how to tackle this problem? :)

Thanks for looking out for that test case! Hmm might be tricky, will does combining these work?: 
- Add a scroll event on the test
- [Using a setTimeout to wait](https://v1.test-utils.vuejs.org/guides/testing-async-components.html#:~:text=%24nextTick%20or-,setTimeout,-to%20ensure%20any) before asserting the button

jovyntls commented on docs/userGuide/syntax/scrollTopButton.md:
```diff
@@ -0,0 +1,39 @@
+| Name     | Type     | Default                 | Description                                                                           |
+|----------|----------|-------------------------|---------------------------------------------------------------------------------------|
+| icon     | `String` | `fa fa-arrow-circle-up` | Icon used for button.                                                                 |
+| iconSize | `String` | ``                      | Size of button. Supports integer-scaling of the icon dimensions e.g. 2x, 3x, 4x, etc. |
```
Could we use kebab-case (`icon-size`) instead of camelCase here? camelCase is a bit of an antipattern in HTML attributes and may cause unexpected behaviour, such as with jQuery, because all attributes will get converted to lowercase ([HTML spec](https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes)).
Also, we can have consistency with our other components like `<box icon-size="..."></box>

yucheng11122017 replied:
Sorry @jovyntls did the comment get cut off?

jovyntls replied:
It did, sorry for the confusion 😅 Meant to say: other components like `Box`

yucheng11122017 replied:
I left the default iconSize to be 'lg' instead of null, otherwise it will be too small. Is that ok?

jovyntls replied:
Yup that sounds good! Thanks @yucheng11122017  :)

jovyntls commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,140 @@
+};
+</script>
+
+<style scoped>
```
```suggestion
<style>
```
In light of #1768  - non-scoped styles have better performance and our class here should be selective enough :)

jovyntls commented on docs/userGuide/syntax/scrollTopButton.md:
```diff
@@ -0,0 +1,39 @@
+
+| Name     | Type     | Default                 | Description                                                                           |
+|----------|----------|-------------------------|---------------------------------------------------------------------------------------|
+| icon     | `String` | `fa fa-arrow-circle-up` | Icon used for button.                                                                 |
```
Is possible to support something similar to the `icon` syntax in the Box component (e.g. `icon=:fas-rocket:"`)? This may be more user-friendly since it'll be consistent with the Box component and how icons are used in Markbind!

jovyntls reviewed (commented):
Thanks for the tests @yucheng11122017 ! Just a few more nits

jovyntls reviewed (approved):
LGTM! Nice work on the tests and the implementation 🎊

tlylt commented:
@yucheng11122017 file conflicts need resolving

tlylt commented on package-lock.json:
```diff
@@ -4389,6 +4389,7 @@
 				"merge-source-map": "^1.1.0",
```
Are we making any dependency update in this PR? How come this is updated?

tlylt commented on packages/vue-components/src/ScrollTopButton.vue:
```diff
@@ -0,0 +1,142 @@
+    :style="iconStyle()"
+    aria-hidden="true"
+    @click="handleScrollTop()">
+    <slot name="icon">
```
Three check warnings above, are they relevant/fixable?

tlylt replied:
@yucheng11122017 any thoughts on the above? (In case you missed it)

yucheng11122017 replied:
Sorry should be fixed in [7072690](https://github.com/MarkBind/markbind/pull/2170/commits/70726909b7f57f3bdd2b8fdd9219827f001e9210)!

It's about the indentation right

tlylt replied:
Thanks, I think it's gone now. Weird that I was still seeing it when I commented earlier...

tlylt reviewed (changes requested):
May I confirm if this means the scroll-top button is no longer a default (i.e. included when someone does markbind init?)

yucheng11122017 commented:
@tlylt I added the scroll top button to the default layout in [packages/core/template/default/_markbind/layouts/default.md](https://github.com/MarkBind/markbind/pull/2170/files#diff-a162f26e57f47d023d8b09a552548f1454f7ec046bb9f4aadc6ecfc532acd61c) so when markbind init is called, it will still be included.

tlylt commented:
> @tlylt I added the scroll top button to the default layout in [packages/core/template/default/_markbind/layouts/default.md](https://github.com/MarkBind/markbind/pull/2170/files#diff-a162f26e57f47d023d8b09a552548f1454f7ec046bb9f4aadc6ecfc532acd61c) so when markbind init is called, it will still be included.

Can we include a line explaining how this can optionally be removed in the docs?

Also, this is indeed breaking change as mentioned in the original issue, will pause the merge until we want to move on to v5

tlylt commented:
Can we also include all details in the proposed commit message? Will need these in the release draft for this major change.

tlylt commented:
https://se-education.org/guides/conventions/git.html

@yucheng11122017 can you kindly help to adjust the commit message format according to the guide, please? Main reason why I think the format is superior is not just for the sake of consistency, but also because it highlights the rationale behind a change, which is crucial context for future devs.

tlylt commented:
Tips on how to fix the conflicts:
- ensure the fork is latest
- ensure local master is in sync with fork master
- checkout to pr branch
- `git merge master`
- accept any changes to conflicts in generated test files
- once the merge is done, `npm run updatetest` to generate latest test files

yucheng11122017 commented:
Thank you for the advice @tlylt! I managed to solve the merge conflicts

yucheng11122017 commented:
> Tips on how to fix the conflicts:
> 
> * ensure the fork is latest
> * ensure local master is in sync with fork master
> * checkout to pr branch
> * `git merge master`
> * accept any (just choose one) changes to conflicts in generated test files, doesn't really matter as we will override them
> * once the merge is done, `npm run updatetest` to generate latest test files

Hi @tlylt, I was thinking about adding these instructions about solving merge conflicts to the DG which talks about updating functional test. Might be helpful for other devs!

tlylt commented:
> > Tips on how to fix the conflicts:
> > 
> > * ensure the fork is latest
> > * ensure local master is in sync with fork master
> > * checkout to pr branch
> > * `git merge master`
> > * accept any (just choose one) changes to conflicts in generated test files, doesn't really matter as we will override them
> > * once the merge is done, `npm run updatetest` to generate latest test files
> 
> Hi @tlylt, I was thinking about adding these instructions about solving merge conflicts to the DG which talks about updating functional test. Might be helpful for other devs!

Sounds good 👍
